flink-optimizer/src/main/java/org/apache/flink/optimizer/traversals/RangePartitionRewriter.java

```diff
@@ -0,0 +1,188 @@
+		for (Channel channel : outgoingChannels) {
+			ShipStrategyType shipStrategy = channel.getShipStrategy();
+			if (shipStrategy == ShipStrategyType.PARTITION_RANGE) {
+				if (channel.getDataDistribution() == null) {
+					removeOutgoingChannels.add(channel);
+					appendOutgoingChannels.addAll(rewriteRangePartitionChannel(channel));
+				}
+			}
+		}
+		outgoingChannels.addAll(appendOutgoingChannels);
+		for (Channel channel : removeOutgoingChannels) {
+			outgoingChannels.remove(channel);
+		}
+	}
+
+	private List<Channel> rewriteRangePartitionChannel(Channel channel) {
+		List<Channel> appendOutgoingChannels = new LinkedList<>();
+		PlanNode sourceNode = channel.getSource();
+		PlanNode targetNode = channel.getTarget();
+		int sourceParallelism = sourceNode.getParallelism();
+		int targetParallelism = targetNode.getParallelism();
+		TypeComparatorFactory<?> comparator = Utils.getShipComparator(channel, this.plan.getOriginalPlan().getExecutionConfig());
+		// 1. Fixed size sample in each partitions.
+		long seed = org.apache.flink.api.java.Utils.RNG.nextLong();
+		int sampleSize = 20 * targetParallelism;
+		SampleInPartition sampleInPartition = new SampleInPartition(false, sampleSize, seed);
+		TypeInformation<?> sourceOutputType = sourceNode.getOptimizerNode().getOperator().getOperatorInfo().getOutputType();
+		TypeInformation<IntermediateSampleData> isdTypeInformation = TypeExtractor.getForClass(IntermediateSampleData.class);
+		UnaryOperatorInformation sipOperatorInformation = new UnaryOperatorInformation(sourceOutputType, isdTypeInformation);
+		MapPartitionOperatorBase sipOperatorBase = new MapPartitionOperatorBase(sampleInPartition, sipOperatorInformation, "Sample in partitions");
+		MapPartitionNode sipNode = new MapPartitionNode(sipOperatorBase);
+		Channel sipChannel = new Channel(sourceNode, TempMode.NONE);
+		sipChannel.setShipStrategy(ShipStrategyType.FORWARD, channel.getDataExchangeMode());
+		SingleInputPlanNode sipPlanNode = new SingleInputPlanNode(sipNode, "SampleInPartition PlanNode", sipChannel, DriverStrategy.MAP_PARTITION);
+		sipPlanNode.setParallelism(sourceParallelism);
+		sipChannel.setTarget(sipPlanNode);
+		appendOutgoingChannels.add(sipChannel);
+		this.plan.getAllNodes().add(sipPlanNode);
+
+		// 2. Fixed size sample in a single coordinator.
+		SampleInCoordinator sampleInCoordinator = new SampleInCoordinator(false, sampleSize, seed);
+		UnaryOperatorInformation sicOperatorInformation = new UnaryOperatorInformation(isdTypeInformation, sourceOutputType);
+		GroupReduceOperatorBase sicOperatorBase = new GroupReduceOperatorBase(sampleInCoordinator, sicOperatorInformation, "Sample in coordinator");
+		GroupReduceNode sicNode = new GroupReduceNode(sicOperatorBase);
+		Channel sicChannel = new Channel(sipPlanNode, TempMode.NONE);
+		sicChannel.setShipStrategy(ShipStrategyType.PARTITION_HASH, channel.getShipStrategyKeys(), channel.getShipStrategySortOrder(), null, channel.getDataExchangeMode());
```

Comment:
The ship strategy can be set to Forward because the DOP of the node is 1. That means that all data is sent to the same node (Forward does only imply local communication if sender DOP = receiver DOP).